W-22103685-graphQI-introspection (#1042)
mulesoft/docs-gateway: gateway/1.13/modules/ROOT/pages/policies-included-graphql-introspection-control.adoc

= GraphQL Introspection Control Policy
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images
:keywords: graphql, introspection, api gateway, flex gateway, gateway, policy

[width="100%", cols="5,15"]
|===
>s| Policy Name | GraphQL Introspection Control

> >s| Summary | Lets you block GraphQL introspection meta-fields `__schema`, `__type`, and `__typename` on incoming requests

Reviewer: ```suggestion
>s| Summary | Blocks or allows requests that  access the `__schema`, `__type`, and `__typename` GraphQL meta fields
```

> >s| Summary | Lets you block GraphQL introspection meta-fields `__schema`, `__type`, and `__typename` on incoming requests

Reviewer: ```suggestion
>s| Summary | Blocks or allows requests that  access the `__schema`, `__type`, and `__typename` GraphQL meta fields
```

>s| Category | Security
>s| First Flex Gateway version available | v1.13.0
.1+>.^s| Returned status codes

> | 403 Forbidden — Queries that include a blocked introspection meta-field. The response follows the standard GraphQL error format, including a message such as `Introspection field '__schema' is not allowed`.

Reviewer: ```suggestion
| 403 Forbidden — The response follows the standard GraphQL error format, including a message such as `Introspection field '__schema' is not allowed`.
```

|===

== Summary

> GraphQL servers include built-in introspection fields that allow clients to discover your API's types and structure. The GraphQL Introspection Control policy inspects every incoming operation and blocks requests that reference any of these restricted fields:

Reviewer: ```suggestion
GraphQL servers include built-in introspection fields that allow clients to discover your API's types and structure. Configure The GraphQL Introspection Control policy inspects incoming operations blocks requests that reference restricted fields.

Use the introspection policy to block requests a combination of these fields:

* `__schema`, used to discover the full API schema.
* `__type`, used to look up details for a specific type by name.
* `__typename`, used to read the concrete type name of an object.

Blocking `__typename` can affect GraphQL client libraries that rely on it for result caching.
```

* `__schema`, used to discover the full API schema.
* `__type`, used to look up details for a specific type by name.
* `__typename`, used to read the concrete type name of an object.

Blocking `__typename` can affect GraphQL client libraries that rely on it for result caching.

You can block specific introspection fields independently. By default, Flex Gateway allows all three fields(`blockSchema`, `blockType`, and `blockTypename` each default to `false`). The policy affects traffic only if you set one or more fields to `true`.

> After Flex Gateway resolves the GraphQL query, the policy evaluates the operation text. Because Flex Gateway doesn't inspect the operation text for persisted queries, these requests pass through without introspection checks.

Reviewer: ```suggestion
After Flex Gateway resolves the GraphQL query, the policy evaluates the operation text. Flex Gateway doesn't inspect the operation text for persisted queries and passes them through without introspection checks.
```

= GraphQL Introspection Control Policy
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images
:keywords: graphql, introspection, api gateway, flex gateway, gateway, policy

[width="100%", cols="5,15"]
|===
>s| Policy Name | GraphQL Introspection Control
>s| Summary | Lets you block GraphQL introspection meta-fields `__schema`, `__type`, and `__typename` on incoming requests
>s| Category | Security
>s| First Flex Gateway version available | v1.13.0
.1+>.^s| Returned status codes

> | 403 - The policy rejects the request when the operation includes a blocked meta-field. The JSON response includes GraphQL-style `errors` with a message such as `Introspection field '__schema' is not allowed`.

Reviewer: ```suggestion
| `403` - The policy rejects the request and includes a header containg the restricted meta fields.
```

|===

== Summary

> GraphQL servers expose built-in introspection meta-fields. Clients use them to discover types and fields. The GraphQL Introspection Control policy inspects each incoming GraphQL operation and blocks requests that reference meta-fields you configure to block:

Reviewer: ```suggestion
The GraphQL Introspection Control policy inspects each incoming GraphQL operation and blocks requests that reference configured GraphQL meta fields. You can allow or block these meta fields:
```

> * `__schema`, used for full schema discovery

Reviewer: ```suggestion
* `__schema`: used for full schema discovery
```

> * `__type`, used to look up a type by name

Reviewer: ```suggestion
* `__type`: used to look up a type by name
```

> * `__typename`, used to read the concrete type name of an object

Reviewer: ```suggestion
* `__typename`: used to read the type name of an object
```

Could maybe rewrite not to say "used for"

> You choose independently whether to block each meta-field. By default, none are blocked, so all three meta-fields are allowed and the policy doesn't change traffic until you set one or more block flags to `true`.

Reviewer: ```suggestion
By default, this policy doesn't block any request.
```

> After Flex Gateway resolves the GraphQL query, the policy evaluates the operation text. For persisted queries, the gateway resolves the document without the usual inline query text, so the policy doesn't inspect those requests, and they pass through without introspection checks.

Reviewer: ```suggestion
After the policy resolves the GraphQL query, the policy evaluates the operation text. For persisted queries, the gateway resolves the document without the inline query text, and the requests pass through without introspection checks.
```

> If a request includes a meta-field you configured to block, the policy returns *403 Forbidden* with `Content-Type: application/json` and a JSON body that includes an `errors` array, such as `{"errors":[{"message":"Introspection field '__schema' is not allowed"}]}`.

Reviewer: ```suggestion
If a request includes a bloack meta field, the policy returns `403` error code with a `Content-Type: application/json` header that includes an  `errors` containing the restricted meta fields the request tried to access. For example, `{"errors":[{"message":"Introspection field '__schema' is not allowed"}]}`.
```

> Use this policy on Flex Gateway API instances that expose a GraphQL interface. Apply it at the API or resource scope as needed.

Reviewer: ```suggestion
```

== Configuring Policy Parameters

include::partial$policy-title-headers.adoc[tag=configFile]

----
- policyRef:
    name: graphql-introspection-control-flex
  config:
    blockSchema: <boolean> // OPTIONAL, default: false
    blockType: <boolean> // OPTIONAL, default: false
    blockTypename: <boolean> // OPTIONAL, default: false
----

[%header%autowidth.spread,cols="a,a,a,a"]
|===

> | Parameter | Required or Optional | Default Value | Description

Reviewer: ```suggestion
| Parameter | Required | Default Value | Description
```

| `blockSchema`

> | Optional

Reviewer: ```suggestion
| No
```

| `false`

> | When `true`, the policy returns 403 when the operation references the `__schema` meta-field anywhere in the document.

Reviewer: ```suggestion
| When `true`, the policy blocks requests that reference the `__schema` meta field.
```

| `blockType`

> | Optional

Reviewer: ```suggestion
| No
```

| `false`

> | When `true`, the policy returns 403 when the operation references the `__type` meta-field.

Reviewer: ```suggestion
| When `true`, the policy blocks requests that reference the `__type` meta field.
```

| `blockTypename`

> | Optional

Reviewer: ```suggestion
| No
```

| `false`

> | When `true`, the policy returns 403 when the operation references the `__typename` meta-field.

Reviewer: ```suggestion
| When `true`, the policy blocks requests that reference the `__typename` meta-field.
```

|===

==== Resource Configuration Example

> The following example blocks schema and type introspection while still letting clients use `__typename` where they need it:

Reviewer: ```suggestion
This example blocks requests that reference `__schema` or `__type`, while allowing requests that reference `__typename`:
```

[source,yaml]
----
- policyRef:
    name: graphql-introspection-control-flex
  config:
    blockSchema: true
    blockType: true
    blockTypename: false
----

include::partial$policy-title-headers.adoc[tag=ui]

[%header%autowidth.spread,cols="a,a,a,a"]
|===
s|Field s|Description s|Default Value s|Required or Optional

| Block `__schema`

> | When enabled, clients can't use the `__schema` meta-field to discover the full API schema.

Reviewer: ```suggestion
| When enabled, the policy blocks requests that reference the `__schema` meta field.
```

| Disabled, schema introspection allowed | Optional

| Block `__type`

> | When enabled, clients can't use the `__type` meta-field to query details about a specific type by name.

Reviewer: ```suggestion
| When enabled, the policy blocks requests that reference the `__type` meta field.
```

| Disabled, type introspection allowed | Optional

| Block `__typename`

> | When enabled, clients can't use the `__typename` meta-field to read the concrete type name of an object.

Reviewer: ```suggestion
| When enabled, the policy blocks requests that reference the `__typename` meta field.
```

| Disabled, `__typename` allowed | Optional

> | Advanced options

Reviewer: Remove advanced options

| Standard policy options such as policy version and application rules.
| |
|===

> [NOTE]

Reviewer: Remove note
